### DISCUSSION TITLE
WebRTC in Bun

### DISCUSSION THREAD

**Discussion by mchepel** [oven-sh/bun#1833] [Ideas]
There are use cases (like P2P via WebRTC) that would make Bun the tool of choice for many projects because of it's simplicity and performance. Node has [node-datachannel](https://github.com/murat-dogan/node-datachannel) that uses [libdatachannel](https://github.com/paullouisageneau/libdatachannel) library.

Bun is using blazing fast [uWebSockets](https://github.com/uNetworking/uWebSockets) networking library. It seems porting or implementing WebRTC feature in Bun would add incredible value to the project and speed up adoption.

**Comment by caramboleyo** [comment]
I tried to use node-datachannel with bun. It seems there is a problem with udp, most of the time it does not work giving you a WebRTC Error but sometimes it suddenly does.

**Comment by ashkalor** [comment]
Yes, working with any realtime low latency data requires webrtc at the minimum to be able to send data over udp. I am unable to use BUN because of this small requirement. Hoping to see updates soon.

**Comment by guest271314** [comment]
Any reason you just don't use `libdatachannel` as a subprocess?

**Comment by ashkalor** [comment]
Are you suggesting I use BUN's FFI to interact with libdatachannel to run it as subprocess? . I believe that will also make the performance better compared to node-datachannel. I haven't worked much on writting wrappers to other languages before maybe thats why I didnt think about it from the top of my mind.  By doing so can I still interface with existing js libraries for signalling?.  Do you have any specific pointers or process for how I should approach this?
Thanks for your help.

**Reply by alfredomariamilano** [reply]
Why not use node-datachannel directly?

**Reply by guest271314** [reply]
You should be able to just use `spawn()` or `node:child_process`. The main goal is just exchanging SDP. I've done that a few different ways

- Stream audio from Firefox browser to Chromium browser, using clipboard to exchange SDP between the applications [Capture monitor device at Nightly stream to Chromium](https://gist.github.com/guest271314/04a539c00926e15905b86d05138c113c)
- libdatachennel copy/pase example https://github.com/paullouisageneau/libdatachannel/issues/544
- Offscreen document <=> WebRTC Data Channel <=> Web page to establish communication between an offscreen Web extension and an arbitrary Web page [offscreen-webrtc](https://github.com/guest271314/offscreen-webrtc)
- Establish WebRTC Data Channel connection between an Isolated Web App and an arbitrary Web page, using SDP in query string [script.js](https://github.com/guest271314/telnet-client/blob/user-defined-tcpsocket-controller-web-api/assets/script.js#L24-L142), and WICG File System Access API [direct-socket-controller.js](https://github.com/guest271314/telnet-client/blob/user-defined-tcpsocket-controller-web-api/direct-sockets/direct-socket-controller.js#L22-L104)

I basically used this pattern, within different contexts, using different methods in the given environments, to exchange SDP

- [A complete example for a WebRTC datachannel with manual signaling](https://stackoverflow.com/a/54985729)

> If you and a friend exchange the offer/answer somehow, you now have a direct peer-to-peer connection. This should work around the world (modulo symmetric [NAT](https://en.wikipedia.org/wiki/Network_address_translation) routers); no data server involved.
